src/sys/windows/selector.rs

```diff
@@ -1,596 +1,545 @@
+    pub(super) fn clone_inner(&self) -> Arc<SelectorInner> {
+        self.inner.clone()
     }
 }
 
-/// Helper struct used for TCP and UDP which bundles a `binding` with a
-/// `SetReadiness` handle.
-pub struct ReadyBinding {
-    binding: Binding,
-    readiness: Option<SetReadiness>,
+#[derive(Debug)]
+pub struct SelectorInner {
+    cp: CompletionPort,
+    active_poll_count: AtomicUsize,
+    update_queue: Mutex<VecDeque<Arc<Mutex<SockState>>>>,
+    afd_group: Mutex<Vec<Arc<Afd>>>,
 }
 
-impl ReadyBinding {
-    /// Creates a new blank binding ready to be inserted into an I/O object.
-    ///
-    /// Won't actually do anything until associated with an `Selector` loop.
-    pub fn new() -> ReadyBinding {
-        ReadyBinding {
-            binding: Binding::new(),
-            readiness: None,
-        }
+impl SelectorInner {
+    pub fn new() -> io::Result<SelectorInner> {
+        CompletionPort::new(0).map(|cp| SelectorInner {
+            cp: cp,
+            active_poll_count: AtomicUsize::new(0),
+            update_queue: Mutex::new(VecDeque::new()),
+            afd_group: Mutex::new(Vec::new()),
+        })
     }
 
-    /// Returns whether this binding has been associated with a selector
-    /// yet.
-    pub fn registered(&self) -> bool {
-        self.readiness.is_some()
-    }
+    pub fn select(&self, events: &mut Events, timeout: Option<Duration>) -> io::Result<()> {
+        events.clear();
 
-    /// Acquires a buffer with at least `size` capacity.
-    ///
-    /// If associated with a selector, this will attempt to pull a buffer from
-    /// that buffer pool. If not associated with a selector, this will allocate
-    /// a fresh buffer.
-    pub fn get_buffer(&self, size: usize) -> Vec<u8> {
-        match self.binding.selector.borrow() {
-            Some(i) => i.buffers.lock().unwrap().get(size),
-            None => Vec::with_capacity(size),
-        }
-    }
+        self.update_sockets_events()?;
 
-    /// Returns a buffer to this binding.
-    ///
-    /// If associated with a selector, this will push the buffer back into the
-    /// selector's pool of buffers. Otherwise this will just drop the buffer.
-    pub fn put_buffer(&self, buf: Vec<u8>) {
-        if let Some(i) = self.binding.selector.borrow() {
-            i.buffers.lock().unwrap().put(buf);
-        }
-    }
+        self.active_poll_count.fetch_add(1, Ordering::SeqCst);
 
-    /// Sets the readiness of this I/O object to a particular `set`.
-    ///
-    /// This is later used to fill out and respond to requests to `poll`. Note
-    /// that this is all implemented through the `SetReadiness` structure in the
-    /// `poll` module.
-    pub fn set_readiness(&self, set: Ready) {
-        if let Some(ref i) = self.readiness {
-            trace!("set readiness to {:?}", set);
-            i.set_readiness(set).expect("event loop disappeared?");
-        }
-    }
+        let mut iocp_events = vec![CompletionStatus::zero(); events.capacity()];
```

Comment:
Ideally we would not allocate a new vector in each call to poll.

Can we use the provided `events: &mut Events` for storage?

Reply:
Done